pkg/registry/apis/provisioning/repository_controller.go

```diff
@@ -215,114 +198,119 @@ func (rc *RepositoryController) processNextWorkItem(ctx context.Context) bool {
-			return fmt.Errorf("failed to get parser for %s: %w", repo.Config().Name, err)
-		}
-
-		id, err := rc.identities.WorkerIdentity(ctx, namespace)
-		if err != nil {
-			return err
-		}
-		ctx = identity.WithRequester(ctx, id)
-
-		replicator, err := jobs.NewSyncer(repo, rc.resourceLister, parser)
-		if err != nil {
-			return fmt.Errorf("error creating replicator")
-		}
-
-		if err := replicator.Unsync(ctx); err != nil {
-			return fmt.Errorf("unsync repository: %w", err)
-		}
-
-		return nil
-	}
-
-	cachedRepo, err := rc.repoLister.Repositories(namespace).Get(name)
+	obj, err := rc.repoLister.Repositories(namespace).Get(name)
 	switch {
 	case apierrors.IsNotFound(err):
 		return errors.New("repository not found in cache")
 	case err != nil:
 		return err
 	}
 
-	id, err := rc.identities.WorkerIdentity(ctx, cachedRepo.Namespace)
+	healthAge := time.Since(time.UnixMilli(obj.Status.Health.Checked))
+	hasSpecChanged := obj.Generation != obj.Status.ObservedGeneration
+	if !hasSpecChanged && obj.DeletionTimestamp == nil && (healthAge < time.Hour*4) {
+		logger.Info("repository spec unchanged")
+		return nil
+	}
+
+	ctx := context.Background()
+	id, err := rc.identities.WorkerIdentity(ctx, namespace)
 	if err != nil {
 		return err
 	}
 	ctx = identity.WithRequester(ctx, id)
-	logger = logger.With("repository", cachedRepo.Name, "namespace", cachedRepo.Namespace)
-	ctx = logging.Context(ctx, logger)
+	logger = logger.WithContext(ctx)
 
-	repo, err := rc.repoGetter.AsRepository(ctx, cachedRepo)
+	repo, err := rc.repoGetter.AsRepository(ctx, obj)
 	if err != nil {
 		return fmt.Errorf("unable to create repository from configuration: %w", err)
 	}
 
-	hasSpecChanged := cachedRepo.Generation != cachedRepo.Status.ObservedGeneration
-	if !hasSpecChanged {
-		logger.Info("repository spec unchanged")
-		return nil
-	}
+	// Safe to edit the repository from here
+	obj = obj.DeepCopy()
+	hooks, _ := repo.(repository.RepositoryHooks)
+	status := &obj.Status
+	var sync *provisioning.SyncOptions
 
-	logger.Info("repository spec changed", "previous", cachedRepo.Status.ObservedGeneration, "current", cachedRepo.Generation)
+	switch {
+	// Delete (note this switch does not fallthrough to the health check)
+	case obj.DeletionTimestamp != nil:
+		logger.Info("handle repository delete")
+		if hooks != nil {
+			// FIXME... this should be converted to a "remove-webhook" finalizer
+			eee := hooks.OnDelete(ctx)
+			if eee != nil {
+				logger.Warn("Error running deletion hooks", "err", err)
+			}
+		}
 
-	var status *provisioning.RepositoryStatus
-	hooks, ok := repo.(repository.RepositoryHooks)
-	if ok {
-		if cachedRepo.Status.ObservedGeneration > 0 {
-			logger.Info("handle repository update")
-			status, err = hooks.OnUpdate(ctx)
-			if err != nil {
-				return fmt.Errorf("handle repository update: %w", err)
+		// Process any finalizers
+		if len(obj.Finalizers) > 0 {
+			for _, finalizer := range obj.Finalizers {
+				switch finalizer {
+				case "cleanup":
```

Comment:
IIUC, we would need a different finalizer for replacing the functionality of `DeletePolicyRetain`, right? (Similar to how K8s uses the `orphan` finalizer, to skip deleting dependent objects.

Reply:
I like the `orphan` naming for that one :)